### PR TITLE
[training migration] Migrate mamba builder

### PR DESCRIPTION
**Pull request by maanug-nv**
# What does this PR do ?
This PR migrates the ModelConfig+ModelBuilder system built in Megatron-Bridge (https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/2241), including the config and builder for Mamba/Hybrid model.
See original PR for more design details.

This should replace the existing `hybrid_builders.py` method of initializing MCore Hybrid models.

#4656 (which is still WIP) demonstrates how this will abstraction will be used and integrated into the setup/initialization phase of the training code.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

**Comment by maanug-nv** [comment]
Remaining tasks for this PR before ready for review:
- [x] Rename Mamba->Hybrid
- [x] Ensure no gaps between `hybrid_builders.hybrid_builder()` and `MambaBuilder.build_model()`
- [x] Ensure no gaps between `unimodal_build_distributed_models()` and `get_model()`
- [x] Integrate MambaConfig+MambaBuilder into `pretrain_hybrid.py` and replace `hybrid_builder()`

**Comment by maanug-nv** [comment]
Will follow-up this PR with migration of `setup()` and `initialize_megatron()` to actually call the builder's build function like so:
```py
builder_cls = model_config.get_builder_cls()
builder = builder_cls(model_config)
return builder.build_distributed_models(
    pg_collection=pg_collection,
    ddp_config=cfg.ddp,
    overlap_param_gather_with_optimizer_step=cfg.optimizer.overlap_param_gather_with_optimizer_step,
    use_megatron_fsdp=cfg.dist.use_megatron_fsdp,
    use_torch_fsdp2=cfg.dist.use_torch_fsdp2,
    data_parallel_random_init=cfg.rng.data_parallel_random_init,
)
```

File: megatron/training/models/base.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+    """Class variable with full path to builder class (e.g.,
+    'megatron.bridge.builders.GPTModelBuilder').
+    """
+
+    # === ModelOpt ===
+    restore_modelopt_state: bool = False
+    """Restore ModelOpt quantization/sparsity state."""
+
+    # === HuggingFace Metadata ===
+    hf_model_id: str | None = None
```
Comment (jaredcasper) [inline megatron/training/models/base.py]: This seems odd... what's the point?

File: megatron/training/models/base.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+    """HuggingFace model identifier."""
+
+    generation_config: Any | None = None
+    """Generation configuration."""
```
Comment (jaredcasper) [inline megatron/training/models/base.py]: What does this mean?

File: megatron/training/models/dist_utils.py [hunk 3]
```diff
@@ -0,0 +1,393 @@
+# Copyright (c) 2026, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (jaredcasper) [inline megatron/training/models/dist_utils.py]: Wrong license, and we don't put license in files.

File: megatron/training/models/hybrid.py [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+    ModelConfig,
+    compose_hooks,
+)
+from megatron.training.models.dist_utils import unimodal_build_distributed_models
+from megatron.training.vocab_utils import calculate_padded_vocab_size
+
+logger = logging.getLogger(__name__)
+
+
+def transformer_engine_hybrid_stack_spec() -> ModuleSpec:
```
Comment (jaredcasper) [inline megatron/training/models/hybrid.py]: Why are these functions? I think it was a mistake to make the spec getters functions in GPT, I'd rather not make that mistake again.

File: megatron/training/models/base.py (hunk 1 above)
Comment (maanug-nv) [reply megatron/training/models/base.py]: In Megatron-Bridge, when we create a model/model config from some HF model source, we set this attribute to the HF model id (eg `nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-BF16`) so we know where it came from and can export back. This is again for the sake of preserving info in the checkpoint. 
So the workflow might look like:
1. Convert an HF model to a Megatron checkpoint
2. Do SFT/RL with that checkpoint
3. Convert back to huggingface format

For the last step, a user would call something like `AutoBridge.export_from_megatron_to_hf("/path/to/dist/checkpoint")` . Having this attribute in the model config allows this to 'just work' and the user doesn't need to specify an HF model ID. Especially internally, when sharing checkpoints, the user doesn't always know which HF repo the megatron checkpoint originally was converted from.

I'm also not sure how to keep only this attribute in Megatron-Bridge.

File: megatron/training/models/dist_utils.py (hunk 3 above)
Comment (maanug-nv) [reply megatron/training/models/dist_utils.py]: My bad, this was a copy paste mistake . thought claude review would catch stuff like this...
fixed in 025e391afac6e44a4856bed8b6bb5164e824d42d

File: megatron/training/models/base.py (hunk 2 above)
Comment (maanug-nv) [reply megatron/training/models/base.py]: This is the HF GenerationConfig. I don't think i can give a full explanation for why it needs to live here, hoping @yaoyu-33 or @cuichenx can do that.
But i'll also update the docstring, since it's totally not clear that this is referring to the HF GenerationConfig

File: megatron/training/models/hybrid.py (hunk 4 above)
Comment (maanug-nv) [reply megatron/training/models/hybrid.py]: I don't recall anymore. I thought it was so that we can reuse logic to serialize/deserialize callables, but I don't see that in the serialization APIs.

I can just drop `transformer_engine_hybrid_stack_spec` and `modelopt_hybrid_stack_spec` and move the conditions in `get_default_hybrid_stack_spec` into `HybridModelBuilder.build_model()` (making `hybrid_stack_spec` in the config `None` by default). Does that make sense to you?

Comment (maanug-nv) [reply megatron/training/models/hybrid.py]: This would make '--spec' no longer a required argument for `pretrain_hybrid.py`

File: megatron/training/models/base.py (hunk 1 above)
Comment (jaredcasper) [reply megatron/training/models/base.py]: Still feels very strange to have a configuration option in Megatron called "hf_model_id". (and also a generation_config that is only for hugging face.)

Maybe can we have a `extra_checkpoint_metadata` arg that the user can fill in with whatever metadata they might want to include in the checkpoint. Then mbridge can use that to encode the hf_model_id and generation_config? (could even be a dict or something to allow non-string but serializable values)

File: megatron/training/models/hybrid.py (hunk 4 above)
Comment (jaredcasper) [reply megatron/training/models/hybrid.py]: yes, I think that makes sense.

File: megatron/training/models/base.py (hunk 1 above)
Comment (maanug-nv) [reply megatron/training/models/base.py]: I think this is fine, 99% sure. I think when @yaoyu-33 is back from ooto, we'll discuss whether the 'hf_model_id' and 'generation_config' are still needed and find exactly where they are being used right now, to confirm if a metadata dict will work.

File: megatron/training/models/base.py (hunk 2 above)
Comment (maanug-nv) [reply megatron/training/models/base.py]: will remove this, no longer needed

File: megatron/training/models/base.py (hunk 1 above)
Comment (maanug-nv) [reply megatron/training/models/base.py]: i think we are in agreement on extra metadata

File: megatron/training/models/hybrid.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+)
+from megatron.training.models.dist_utils import unimodal_build_distributed_models
+from megatron.training.vocab_utils import calculate_padded_vocab_size
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(kw_only=True)
+class HybridModelConfig(ModelConfig):
+    """Configuration for a Megatron Core Hybrid (SSM) model.
```
Comment (jaredcasper) [inline megatron/training/models/hybrid.py]: I think drop the (SSM), it's not tied to that.

File: megatron/training/models/hybrid.py [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+            super().__setattr__(name, value)
+
+    def finalize(self) -> None:
+        """One time validation to run once config is ready to be used by builder."""
+
+        if hasattr(self.transformer, "finalize") and callable(self.transformer.finalize):
+            self.transformer.finalize()
+
+
+class HybridModelBuilder(ModelBuilder[HybridModel, HybridModelConfig]):
```
Comment (jaredcasper) [inline megatron/training/models/hybrid.py:113]: Why does this inherit from HybridModelConfig? Shouldn't it just store the config?

Comment (maanug-nv) [reply megatron/training/models/hybrid.py:113]: It doesn't. `ModelBuilder` is `Generic[ModelT, BuildConfigT]`, so `HybridModelBuilder[HybridModel, HybridModelConfig]` is a parameterization of the generic base, not inheritance from `HybridModelConfig`. The config is stored on `self._model_config` by the base `__init__`.

Comment (maanug-nv) [reply megatron/training/models/hybrid.py:113]: https://github.com/NVIDIA/Megatron-LM/pull/4550/changes#diff-9b17161d4e7e59c04e1b06bc223aba038386037f3634c7b870dda3c8fd719305R168

**Review by yaoyu-33 — APPROVED** [review]

File: megatron/training/models/hybrid.py (hunk 6 above)
Comment (jaredcasper) [reply megatron/training/models/hybrid.py:113]: ah, sorry, didn't read close enough.

**Review by jaredcasper — APPROVED** [review]
